### PR TITLE
Adding support for tab expansion

### PR DESCRIPTION
**Pull request by danieljamespost**
Tab expansion can be activated for a given buffer by executing `Tabexpand`. This will add the number of spaces set by the Tab command in place of '\t' when the tab key is pressed. This option can also be passed as an environment variable on startup, similar to "tabstop" in the form of `tabexpand=true exec edwood`. 

The number of spaces inserted is configurable. For example, if you change the tabwidth from 4 to 8 by executing `Tab 8`, this will also replace the 4 spaces used for indentation with 8 spaces if Tabexpand is activated.

### CODE REVIEW

File: dat.go [hunk 1]
```diff
@@ -75,6 +75,7 @@ var (
 	globalincref bool
 	seq          int
 	maxtab       uint // size of a tab, in units of the '0' character
+	texpand      bool // defines whether to expand tab to spaces
```
Comment (rjkroege) [inline dat.go]: Nit: can you call this tabexpand? It will make easier for (at least me) to remember what it's for later.

File: exec.go [hunk 2]
```diff
@@ -703,12 +705,31 @@ func tab(et *Text, _ *Text, argt *Text, _, _ bool, arg string) {
 		args := strings.Split(arg, " ")
 		arg = args[0]
 		p := arg
+		if len(p) == 0 {
```
Comment (rjkroege) [inline exec.go]: I think that there is a much better general fix here. Can you file an issue and add a `TODO(rjk):` to use a tokenizer on the results of `getarg`. Seems like a nice opportunity to refactor.

File: exec.go [hunk 3]
```diff
@@ -703,12 +705,31 @@ func tab(et *Text, _ *Text, argt *Text, _, _ bool, arg string) {
...
+			return
+		}
 		if '0' <= p[0] && p[0] <= '9' {
 			tab, _ = strconv.ParseInt(p, 10, 16)
 		}
 	}
 	if tab > 0 {
 		if w.body.tabstop != int(tab) {
+
```
Comment (rjkroege) [inline exec.go]: nit: unnecessary blank line.

File: exec.go [hunk 4]
```diff
@@ -703,12 +705,31 @@ func tab(et *Text, _ *Text, argt *Text, _, _ bool, arg string) {
...
+			// replace spaces with new tab width
+			if w.body.texpand {
```
Comment (rjkroege) [inline exec.go]: This code updates the existing text? Just using an Edit command would seem to be much more the Acme way?

File: internal/frame/frameopts.go [hunk 5]
```diff
@@ -72,5 +72,5 @@ func (ctx *optioncontext) computemaxtab(maxtab, ftw int) int {
 	if ctx.maxtabchars < 0 {
 		return maxtab
 	}
-	return ctx.maxtabchars * ftw
+	return ctx.maxtabchars * (ftw / 2)
```
Comment (rjkroege) [inline internal/frame/frameopts.go]: Why did you need to do this?

File: text.go [hunk 6]
```diff
@@ -65,6 +65,7 @@ type Text struct {
 	q1      int
 	what    TextKind
 	tabstop int
+	texpand bool
```
Comment (rjkroege) [inline text.go]: It seems like you have the flag on `wind` and `text`? Why?

File: wind.go [hunk 7]
```diff
@@ -66,7 +67,7 @@ func (w *Window) initHeadless(clone *Window) *Window {
 	w.tag.w = w
 	w.taglines = 1
 	w.tagsafe = false
-	w.tagexpand = true
+	w.tagexpand = false
```
Comment (rjkroege) [inline wind.go]: Why this change? It seems unrelated to your CL and modifies the UI? Maybe you meant `w.tabexpand = false`?

File: go.mod [hunk 8]
```diff
@@ -3,7 +3,7 @@ module github.com/rjkroege/edwood
 go 1.11
 
 require (
-	9fans.net/go v0.0.0-20181112161441-237454027057
+	9fans.net/go v0.0.2
```
Comment (rjkroege) [inline go.mod]: can you revert this or make it a different CL

File: text.go (hunk 6 above)
Comment (rjkroege) [inline text.go]: I'd like this to be called tabexpand.

File: dat.go (hunk 1 above)
Comment (danieljamespost) [reply dat.go]: That makes sense. Just updated my PR with this change.

File: exec.go (hunk 2 above)
Comment (danieljamespost) [reply exec.go]: Updated my PR with the comment and created the issue.

File: exec.go (hunk 3 above)
Comment (danieljamespost) [reply exec.go]: Fixed in a later commit.

File: exec.go (hunk 4 above)
Comment (danieljamespost) [reply exec.go]: You're right, that would be more idiomatic. My thought process here was to replicate the behavior of executing `Tab <width>` when tab expansion is off by re-indenting automatically when tab expansion is activated. In any case, I've updated my PR to remove this unless you think there is a more elegant way of handling it.

File: internal/frame/frameopts.go (hunk 5 above)
Comment (danieljamespost) [reply internal/frame/frameopts.go]: This was a mistake that I fixed in a later commit.

File: wind.go (hunk 7 above)
Comment (danieljamespost) [reply wind.go]: You're correct, I fixed this in a later commit.

File: go.mod (hunk 8 above)
Comment (danieljamespost) [reply go.mod]: Done.

File: text.go (hunk 6 above)
Comment (danieljamespost) [reply text.go]: Done. Tab expansion isn't persisted in the dump file, but if this is desired I can add that as well. I should have some time this weekend to add tests for these features.

**Comment by fhs** [comment]
I think the changes in the font files are just line ending changes? They should probably be marked as binary in the `.gitattributes` file. I'm also encountering this:
```diff
warning: CRLF will be replaced by LF in build/font/lucm/cyrillic.9.
The file will have its original line endings in your working directory
warning: CRLF will be replaced by LF in build/font/shinonome/k14.5c01.
The file will have its original line endings in your working directory
diff --git a/build/font/lucm/cyrillic.9 b/build/font/lucm/cyrillic.9
index ac440ca..452da53 100644
Binary files a/build/font/lucm/cyrillic.9 and b/build/font/lucm/cyrillic.9 differ
diff --git a/build/font/shinonome/k14.5c01 b/build/font/shinonome/k14.5c01
index e815e9f..2b89054 100644
Binary files a/build/font/shinonome/k14.5c01 and b/build/font/shinonome/k14.5c01 differ
````

File: exec.go (hunk 2 above)
Comment (rjkroege) [reply exec.go]: thanks

**Comment by rjkroege** [comment]
Thanks for all the work on here. This is cool.
